Makefile

```diff
@@ -36,111 +33,214 @@ TESTSRC := $(COMMON_TESTSRC) $(FUNCTIONAL_TESTSRC) $(UNIT_TESTSRC)
+#*************************************************************#
+# Flags to select release vs debug builds, verbosity, etc.
 #
-# Note(s):
-#  - Address sanitizer builds: -fsanitize=address
-#     - Ctests will be silently skipped with clang builds. (Known issue.)
-#       Use gcc to build in Asan mode to run unit-tests.
-#     - Tests will run slow in address sanitizer builds.
+
+help::
+	@echo Environment variables controlling the build
+	@echo '  BUILD_DIR: Base name for build output (Default: "build").'
+	@echo '    $$(BUILD_DIR)/obj: object files'
+	@echo '    $$(BUILD_DIR)/lib: libraries'
+	@echo '    $$(BUILD_DIR)/bin: executables'
+	@echo '  Note: setting BUILD_MODE and other flags may further modify BUILD_DIR'
+	@echo
+
+ifndef BUILD_DIR
+BUILD_DIR := build
+endif
+
 #
-#  - Memory sanitizer builds: -fsanitize=memory
-#     - Builds will fail with gcc due to compiler error. Use clang instead.
-#     - Tests will run even slower in memory sanitizer builds.
+# Build mode
 #
-CFLAGS += $(DEFAULT_CFLAGS) -Ofast -flto
-LDFLAGS += $(DEFAULT_LDFLAGS) -Ofast -flto
-LIBS = -lm -lpthread -laio -lxxhash $(LIBCONFIG_LIBS)
+ifndef BUILD_MODE
+BUILD_MODE=release
+endif
 
-DEPFLAGS = -MMD -MT $@ -MP -MF $(OBJDIR)/$*.d
+ifeq "$(BUILD_MODE)" "debug"
+CFLAGS  += -DSPLINTER_DEBUG
+#LDFLAGS +=
+BUILD_DIR:=$(BUILD_DIR)-debug
+else ifeq "$(BUILD_MODE)" "release"
+CFLAGS   += -Ofast -flto
+LDFLAGS  += -Ofast -flto
+else ifeq "$(BUILD_MODE)" "optimized-debug"
+CFLAGS  += -DSPLINTER_DEBUG
+CFLAGS  += -Ofast -flto
+LDFLAGS += -Ofast -flto
+BUILD_DIR := $(BUILD_DIR)-optimized-debug
+else
+$(error Unknown BUILD_MODE "$(BUILD_MODE)".  Valid options are "debug", "optimized-debug", and "release".  Default is "release")
+endif
 
-COMPILE.c = $(CC) $(DEPFLAGS) $(CFLAGS) $(INCLUDE) $(TARGET_ARCH) -c
+help::
+	@echo '  BUILD_MODE: "release", "debug", or "optimized-debug" (Default: "release")'
 
-####################################################################
-# The main targets
 #
+# address sanitizer
+#
+ifndef BUILD_ASAN
+BUILD_ASAN=0
+endif
 
-all: libs tests $(EXTRA_TARGETS)
+ifeq "$(BUILD_ASAN)" "1"
+CFLAGS  += -fsanitize=address
+LDFLAGS += -fsanitize=address
+BUILD_DIR:=$(BUILD_DIR)-asan
+else ifeq "$(BUILD_ASAN)" "0"
+else
+$(error Unknown BUILD_ASAN mode "$(BUILD_ASAN)".  Valid values are "0" or "1". Default is "0")
+endif
 
-libs: $(LIBDIR)/libsplinterdb.so $(LIBDIR)/libsplinterdb.a
+help::
+	@echo '  BUILD_ASAN={0,1}: Disable/enable address-sanitizer (Default: disabled)'
 
-tests: $(BINDIR)/driver_test $(BINDIR)/unit_test $(UNIT_TESTBINS)
+#
+# memory sanitizer
+#
+ifndef BUILD_MSAN
+BUILD_MSAN=0
+endif
 
-#*************************************************************#
-# Targets to track whether we have a release or debug build
-release: .release all
-	rm -f .debug .debug-log
+ifeq "$(BUILD_MSAN)" "1"
+CFLAGS  += -fsanitize=memory
+LDFLAGS += -fsanitize=memory
+BUILD_DIR:=$(BUILD_DIR)-msan
+else ifeq "$(BUILD_MSAN)" "0"
+else
```

Comment:
Seems like this can be merged with prev line using `else ifneq "$(BUILD_MSAN)" "0"` --- for readability.